src/model/File_data.h

```diff
@@ -35,20 +35,24 @@ class File_data
             : file_meta_(file_meta),
               contents_(contents) {}
 
-    const dbo::ptr<File_meta>& file_meta() const { return file_meta_; }
-    const Bytes& contents() const { return contents_; }
+	File_data(const dbo::ptr<File_meta>& file_meta)
+			: file_meta_(file_meta) {}
 
+	const dbo::ptr<File_meta>& file_meta() const { return file_meta_; }
+	const Bytes& contents() const { return contents_; }
+	int write_and_commit();
+	int delete_and_commit();
+	int populate_contents();
+	
 private:
-    dbo::ptr<File_meta> file_meta_;
+	dbo::ptr<File_meta> file_meta_;
     Bytes contents_;
+	git_repository* repo_init_(std::string repo_path);
+	bool repo_add_commit_(git_repository *repo, const char* repo_path, 
+					const char* committer, int update = 0);
+	static std::unordered_map<std::string, std::mutex*> mtx_map_;
```

Comment:
Looks like you are using mutexes for concurrency control. Mutexes won’t work because GSC runs in multiple OS processes, and there may be instances of the same `File_data` object in more than one. The way to do this kind of interprocess locking is with a file lock. Check out `man 2 flock`.

Also, oh, I see now why the `new` earlier doesn’t *really* leak—because the map is static, so it’s never getting deallocated anyway. So you are you using a static `std::unordered_map` to associate mutexes with repositories? And then the global lock on the next line is protecting that map? Eek! You really do need `flock` here.

Reply:
I sent you an email about this.